test/test_udacity.py

```diff
@@ -0,0 +1,836 @@
+            alices_loss = ((alices_pred - alices_target) ** 2).sum()
+            alices_loss.backward()
+
+            alices_opt.step()
+            alices_loss = alices_loss.get().data
+
+        alices_model.move(secure_worker)
+        bobs_model.move(secure_worker)
+
+        with th.no_grad():
+
+            model.weight.set_(((alices_model.weight.data + bobs_model.weight.data) / 2).get())
+            model.bias.set_(((alices_model.bias.data + bobs_model.bias.data) / 2).get())
+
+        print("Bob:" + str(bobs_loss) + " Alice:" + str(alices_loss))
+
+    preds = model(data)
+    loss = ((preds - target) ** 2).sum()
+
+    print(preds)
+    print(target)
+    print(loss.data)
+
+    x = 5
+
+    bob_x_share = 2
+    alice_x_share = 3
+
+    decrypted_x = bob_x_share + alice_x_share
+    decrypted_x
+
+    bob_x_share = 2 * 2
+    alice_x_share = 3 * 2
+
+    decrypted_x = bob_x_share + alice_x_share
+    decrypted_x
+
+    # encrypted "5"
+    bob_x_share = 2
+    alice_x_share = 3
+
+    # encrypted "7"
+    bob_y_share = 5
+    alice_y_share = 2
+
+    # encrypted 5 + 7
+    bob_z_share = bob_x_share + bob_y_share
+    alice_z_share = alice_x_share + alice_y_share
+
+    decrypted_z = bob_z_share + alice_z_share
+    decrypted_z
+
+    x = 5
+
+    Q = 23740629843760239486723
+
+    bob_x_share = 23552870267  # <- a random number
+    alice_x_share = Q - bob_x_share + x
+    alice_x_share
+
+    (bob_x_share + alice_x_share) % Q
+
+    x_share = (2, 5, 7)
+
+    import random
+
+    Q = 23740629843760239486723
+
+    def encrypt(x, n_share=3):
+
+        shares = list()
+
+        for i in range(n_share - 1):
+            shares.append(random.randint(0, Q))
+
+        shares.append(Q - (sum(shares) % Q) + x)
+
+        return tuple(shares)
+
+    def decrypt(shares):
+        return sum(shares) % Q
+
+    shares = encrypt(3)
+    shares
+
+    decrypt(shares)
+
+    def add(a, b):
+        c = list()
+        for i in range(len(a)):
+            c.append((a[i] + b[i]) % Q)
+        return tuple(c)
+
+    x = encrypt(5)
+    y = encrypt(7)
+    z = add(x, y)
+    decrypt(z)
+
+    BASE = 10
+    PRECISION = 4
+
+    def encode(x):
+        return int((x * (BASE ** PRECISION)) % Q)
+
+    def decode(x):
+        return (x if x <= Q / 2 else x - Q) / BASE ** PRECISION
+
+    encode(3.5)
+
+    decode(35000)
+
+    x = encrypt(encode(5.5))
+    y = encrypt(encode(2.3))
+    z = add(x, y)
+    decode(decrypt(z))
+
+    bob = bob.clear_objects()
+    alice = alice.clear_objects()
+    secure_worker = secure_worker.clear_objects()
+
+    x = th.tensor([1, 2, 3, 4, 5])
+
+    x = x.share(bob, alice, secure_worker)
+
+    bob._objects
+
+    y = x + x
+
+    y
+
+    y.get()
+
+    x = th.tensor([0.1, 0.2, 0.3])
+
+    x = x.fix_prec()
+
+    x.child.child
+
+    y = x + x
+
+    y = y.float_prec()
+    y
+
+    x = th.tensor([0.1, 0.2, 0.3])
+
+    x = x.fix_prec().share(bob, alice, secure_worker)
+
+    y = x + x
+
+    y.get().float_prec()
+
```

Comment:
Delete these blank lines

Reply:
they're separate cells in the notebook - at some point i'm going to have to come back and find bits of code to change,. the lines help somee